Add guidelines for writing controllers (#4110)
MetaMask/core: docs/writing-controllers.md

## Understand the purpose of the controller pattern

Controllers are foundational pieces within MetaMask's architecture:

- They keep and manage wallet-centric data, which can be saved and recalled upon reloading (accounts, transactions, preferences, etc.).
- They contain business logic that powers functionality in the product (what happens when the wallet is locked, the network is switched, a transaction is completed, etc.).
- They act as a communication layer between service layers (blockchains, internal or external APIs, etc.).
- They allow the application to be divided into logical modules which can be maintained by different teams within the company.

## Use the latest version of `BaseController` for controllers

All controllers should inherit from `BaseController` from the `@metamask/base-controller` package. This provides a few benefits:

- It defines a standard interface for all controllers.
- It introduces the messenger system, which is useful for interacting with other controllers without requiring direct access to them.
- It enforces that `update` is the only way to modify the state of the controller and provides a way to listen for state updates via the messenger.

> - It simplifies initialization by consolidating constructor arguments into one.

Reviewer: nit: for clarity
```suggestion
- It simplifies initialization by consolidating constructor arguments into one options object.
```

Author: Makes sense!

## Don't use `BaseController` for non-controllers

One of the uniquely identifying features of a controller is the ability to manage state.

If you have a class that does not capture any data in state, then your class does not need to inherit from `BaseController` (even if it uses the messaging system).

## Maintain a clear and concise API

The name of the controller should reflect its responsibility. If, when creating a controller, it is difficult to come up with a good name, consider defining the responsibility first.

Each public method and each state property of a controller should have a purpose, and the name of the method or state property should be readable and should reflect the purpose clearly. If something does not need to be public, it should be made private; if it is unnecessary, it should be removed.

> ## Accept a optional, partial representation of state

Reviewer: typo:
```suggestion
## Accept an optional, partial representation of state
```

Author: Good catch!

Although `BaseController` requires a full representation of controller state, in practice, controllers should accept a partial version and then supply missing properties with defaults. In fact, the `state` argument should be optional:

```typescript
type FooControllerState = {
  // ...
};

function getDefaultFooControllerState(): FooControllerState {
  // ...
}

class FooController extends BaseController</* ... */> {
  constructor({
    messenger,
    state = {},
  }: {
    messenger: FooControllerMessenger;
    state: Partial<FooControllerState>;
  }) {
    super({
      // ...
      messenger,
      state: { ...getDefaultFooControllerState(), ...state },
    });

    // ...
  }
}
```

## Provide a default representation of state

Each controller needs a default representation in order to fully initialize itself when [receiving a partial representation of state](#accept-a-partial-representation-of-state). A default representation of state is also useful when testing interactions with a controller's `*:stateChange` event.

> It is customary to export a function which returns this default representation:

Reviewer: nit: Providing an explanation for why we prefer a function might make it easier to encourage this practice.

```suggestion
A function which returns this default representation should be defined and exported. 

Note that the default representation object is not exported directly to ensure that we are always providing a newly initialized instance with a new object reference.
```

Author: Good idea!

});
rootMessenger.subscribe('FooController:someEvent', () => {
  // do something with the event
});
```

## Define the `*:getState` action using the `ControllerGetStateAction` utility type

Each controller needs a type for its `*:getState` action. The `ControllerGetStateAction` utility type from the `@metamask/base-controller` package should be used to define this type.

The name of this type should be `${ControllerName}GetStateAction`.

```typescript
import type { ControllerGetStateAction } from '@metamask/base-controller';

type FooControllerGetStateAction = ControllerGetStateAction<
  'FooController',
  FooControllerState
>;
```

## Define the `*:stateChange` event using the `ControllerStateChangeEvent` utility type

Each controller needs a type for its `*:stateChange` event. The `ControllerStateChangeEvent` utility type from the `@metamask/base-controller` package should be used to define this type.

The name of this type should be `${ControllerName}StateChangeEvent`.

```typescript
import type { ControllerStateChangeEvent } from '@metamask/base-controller';

type FooControllerStateChangeEvent = ControllerStateChangeEvent<
  'FooController',
  FooControllerState
>;
```

> ## Use standard naming for action identifiers
> 
> An action is equivalent to a function, and its identifier should be named in a similar fashion.

Reviewer: nit: for clarity

```suggestion
## Use standard naming scheme for action identifiers

An action is equivalent to a function, and its identifier should be named in "camelCase".
```

Author: Good change.

```typescript
const name = 'FooController';

export type FooControllerShowNotificationAction = {
  type: `${typeof name}:showNotification`;
  handler: () => void;
};
```

> ## Use standard naming for action types
> 
> Types for messenger actions should follow this naming:

Reviewer: nit: "naming scheme" seems like the clearer wording.

```suggestion
## Use standard naming scheme for action types

Types for messenger actions should follow this naming scheme:
```

Author: Makes sense.

  type: `${typeof name}:receiveMessage`;
  payload: () => void;
};
```

🚫 **The event identifier includes "on"**

```typescript
const name = 'FooController';

export type FooControllerOnMessageReceivedEvent = {
  type: `${typeof name}:onMessageReceived`;
  payload: () => void;
};
```

✅ **The event identifier sounds like a statement and does not include "on"**

```typescript
const name = 'FooController';

export type FooControllerMessageReceivedEvent = {
  type: `${typeof name}:messageReceived`;
  payload: () => void;
};
```

## Use standard naming for event types

Types for messenger events should follow this naming:

```
Controller name + Event identifier without controller name prefix + "Event"
```

🚫 **Type name does not include controller name and does not end with "Action"**

```typescript
const name = 'FooController';

export type MessageReceived = {
  type: `${typeof name}:messageReceived`;
  payload: () => void;
};
```

✅ **Type name now matches controller name + event identifier + "Event"**

```typescript
const name = 'FooController';

export type FooControllerMessageReceivedEvent = {
  type: `${typeof name}:messageReceived`;
  payload: () => void;
};
```

> ## Define and export internal action types as one

Reviewer: The meaning of "as one" seems a little ambiguous.

```suggestion
## Define and export a type union for internal action types
```

Author: Makes sense.

```typescript
/* === packages/accounts-controller/src/AccountsController.ts === */

type Account = {
  address: Hex;
}

type AccountsControllerState = {
  selectedAccount: Account;
}

class AccountsController extends BaseController<
  'AccountsController',
  AccountsControllerState,
  // ...
> {
  // ...
}

/* === packages/preferences-controller/src/PreferencesController.ts === */

type PreferencesControllerState = {
  selectedAccountAddress: Hex;
}

class PreferencesController extends BaseController<
  'PreferencesController',
  PreferencesControllerState,
  ...
> {
  // ...
}
```

✅ **Only one controller owns the selected account, and the other consumes it**

```typescript
/* === packages/accounts-controller/src/AccountsController.ts === */

type Account = {
  address: Hex;
};

type AccountsControllerState = {
  selectedAccount: Account;
};

export type AccountsControllerGetStateAction = ControllerGetStateAction<
  'AccountsController',
  AccountsControllerState
>;

class AccountsController extends BaseController<
  'AccountsController',
  AccountsControllerState
  // ...
> {
  // ...
}

/* === packages/preferences-controller/src/PreferencesController.ts === */

> import { AccountsControllerState } from '@metamask/accounts-controller';
> 
> type AllowedActions = AccountsControllerGetState;

Reviewer: Looks like an error?

```suggestion
import { AccountsControllerGetStateAction } from '@metamask/accounts-controller';

type AllowedActions = AccountsControllerGetStateAction;
```

Author: Yes thank you!

An action is equivalent to a function, and its identifier should be named in "camelCase".

```typescript
const name = 'FooController';

export type FooControllerShowNotificationAction = {
  type: `${typeof name}:showNotification`;
  handler: () => void;
};
```

## Use standard naming scheme for action types

Types for messenger actions should follow this naming scheme:

```
Controller name + Action identifier without controller name prefix + "Action"
```

🚫 **Type name does not include controller name and does not end with "Action"**

```typescript
const name = 'FooController';

export type ShowNotification = {
  type: `${typeof name}:show`;
  handler: () => void;
};
```

🚫 **Type name includes controller name, but does not match action identifier minus prefix**

```typescript
const name = 'FooController';

export type FooControllerShowNotificationAction = {
  type: `${typeof name}:show`;
  handler: () => void;
};
```

✅ **Type name now matches controller name + action identifier + "Action"**

```typescript
const name = 'FooController';

export type FooControllerShowAction = {
  type: `${typeof name}:show`;
  handler: () => void;
};
```

> ## Use standard naming for event identifiers

Reviewer: For consistency with https://github.com/MetaMask/core/pull/4110#discussion_r1670855179

```suggestion
## Use standard naming scheme for event identifiers
```

An event is not an action, so the identifier for an event should not describe something that the controller needs to do, but rather something that has has happened to the controller.

You may find it helpful to pretend that the word "on" precedes the identifier (but do not include this in the identifier itself).

🚫 **`receiveMessage` sounds like a command**

```typescript
const name = 'FooController';

export type FooControllerReceiveMessageEvent = {
  type: `${typeof name}:receiveMessage`;
  payload: () => void;
};
```

🚫 **The event identifier includes "on"**

```typescript
const name = 'FooController';

export type FooControllerOnMessageReceivedEvent = {
  type: `${typeof name}:onMessageReceived`;
  payload: () => void;
};
```

✅ **The event identifier sounds like a statement and does not include "on"**

```typescript
const name = 'FooController';

export type FooControllerMessageReceivedEvent = {
  type: `${typeof name}:messageReceived`;
  payload: () => void;
};
```

> ## Use standard naming for event types

Reviewer: For consistency with https://github.com/MetaMask/core/pull/4110#discussion_r1670855179

```suggestion
## Use standard naming scheme for event types
```

};
```

## Define and export a type union for internal action types

A controller should define and export a type union that holds all of its actions. This type union makes it easy to define the type for the controller's messenger.

The name of this type should be `${ControllerName}Actions`.

This type should be only passed to `RestrictedControllerMessenger` as the 2nd type parameter. It should _not_ be included in its 4th type parameter, as that is is used for external events.

🚫 **`FooController['type']` is passed as the 4th type parameter**

```typescript
export type FooControllerActions =
  | FooControllerUpdateCurrencyAction
  | FooControllerUpdateRatesAction;

export type FooControllerMessenger = RestrictedControllerMessenger<
  'FooController',
  FooControllerActions,
  never,
  FooControllerActions['type'],
  never
>;
```

✅ **`never` is passed as the 5th type parameter (assuming no external events)**

```typescript
export type FooControllerActions =
  | FooControllerUpdateCurrencyAction
  | FooControllerUpdateRatesAction;

export type FooControllerMessenger = RestrictedControllerMessenger<
  'FooController',
  FooControllerActions,
  never,
  never,
  never
>;
```

> ## Define and export internal event types as one

Reviewer: For consistency with https://github.com/MetaMask/core/pull/4110#discussion_r1670856411

```suggestion
## Define and export a type union for internal event types
```

A controller should define and export a type union that holds all of its events. This type union makes it easy to define the type for the controller's messenger.

The name of this type should be `${ControllerName}Events`.

This type should be only passed to `RestrictedControllerMessenger` as the 3rd type parameter. It should _not_ be included in its 5th type parameter, as that is is used for external events.

🚫 **`FooControllerEvents['type']` is passed as the 5th type parameter**

```typescript
export type FooControllerEvents =
  | FooControllerMessageReceivedEvent
  | FooControllerNotificationAddedEvent;

export type FooControllerMessenger = RestrictedControllerMessenger<
  'FooController',
  never,
  FooControllerEvents,
  never,
  FooControllerEvents['type']
>;
```

✅ **`never` is passed as the 5th type parameter (assuming no external events)**

```typescript
export type FooControllerEvents =
  | FooControllerMessageReceivedEvent
  | FooControllerNotificationAddedEvent;

export type FooControllerMessenger = RestrictedControllerMessenger<
  'FooController',
  never,
  FooControllerEvents,
  never,
  never
>;
```

> ## Define external action types as one, but do not export them

Reviewer: For consistency with https://github.com/MetaMask/core/pull/4110#discussion_r1670856411

```suggestion
## Define, but do not export, a type union for external action types
```

  | BarControllerDoSomethingAction
  | BarControllerDoSomethingElseAction;

export type FooControllerMessenger = RestrictedControllerMessenger<
  'FooController',
  AllowedActions,
  never,
  AllowedActions['type'],
  never
>;

/* === packages/foo-controller/src/index.ts === */

export type { AllowedActions } from '@metamask/foo-controller';
```

✅ **`AllowedActions['type']` is passed as the 4th type parameter, and `AllowedActions` is _not_ exported**

```typescript
type AllowedActions =
  | BarControllerDoSomethingAction
  | BarControllerDoSomethingElseAction;

export type FooControllerMessenger = RestrictedControllerMessenger<
  'FooController',
  AllowedActions,
  never,
  AllowedActions['type'],
  never
>;
```

If, in a test, you need to access all of the actions included in a controller's messenger allowlist, use the [`ExtractAvailableAction` utility type](../packages/base-controller/tests/helpers.ts):

```typescript
// NOTE: You may need to adjust the path depending on where you are
import { ExtractAvailableAction } from '../../base-controller/tests/helpers';

const messenger = new ControllerMessenger<
  ExtractAvailableAction<FooControllerMessenger>,
  never
>();
```

> ## Define external event types as one, but do not export them

Reviewer: For consistency with https://github.com/MetaMask/core/pull/4110#discussion_r1670856411

```suggestion
## Define, but do not export, a type union for external event types
```

  // ...
}

for (const account of fooController.inactiveAccounts) {
  // ...
}
```

✅

```typescript
class FooController extends BaseController {}

// Then, later...
const fooController = new FooController();

const activeAccounts = fooController.state.accounts.filter(
  (account) => account.isActive,
);
for (const account of activeAccounts) {
  // ...
}

const inactiveAccounts = fooController.state.accounts.filter(
  (account) => !account.isActive,
);
for (const account of inactiveAccounts) {
  // ...
}
```

Instead of getters, try selectors:

💡

```typescript
// === foo-controller ===

type FooControllerState = {
  // ...
};

export fooControllerSelectors = {
  activeAccounts: (state: FooControllerState) => {
    return state.accounts.filter(
      (account) => account.isActive,
    )
  },
  inactiveAccounts: (state: FooControllerState) => {
    return state.accounts.filter(
      (account) => !account.isActive,
    )
  },
};

export class FooController extends BaseController {
}

// === Client ==

import { select } from "@metamask/base-controller";
import { FooController, fooControllerSelectors } from "@metamask/foo-controller";

const fooController = new FooController();

> const [activeAccounts, inactiveAccounts] = select(

Reviewer 2: It should probably be `fooController.state` one line down from here, rather than `fooController`. We'll need to select from the state. But aside from that, this looks basically like I had imagined as well!

> const [activeAccounts, inactiveAccounts] = select(

Author: I ended up updating this example to use `reselect`.
